README.md
# mapbox-cli-py

[![Build Status](https://travis-ci.org/mapbox/mapbox-cli-py.svg?branch=master)](https://travis-ci.org/mapbox/mapbox-cli-py) [![Coverage Status](https://coveralls.io/repos/mapbox/mapbox-cli-py/badge.svg?branch=master&service=github)](https://coveralls.io/github/mapbox/mapbox-cli-py?branch=master)

Command line interface to Mapbox Web Services based on https://github.com/mapbox/mapbox-sdk-py.

## Installation

**For users on OS X**, we recommend installing with [homebrew](http://brew.sh/)
```
$ brew install mapbox/cli/mapbox
```

**For users familiar with Python** and who already have [`pip`](https://pip.pypa.io/en/stable/installing/) installed on their system, you can create a [virtual environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/) and install with
```
(venv)$ pip install mapboxcli
```

Installing locally without a virtual environment using
```
$ pip install --user mapboxcli
```
You'll then need to include ~/.local/bin in your $PATH.

Installing globally is *not recommended* but some users may want to do so under certain circumstances
```
$ sudo pip install mapboxcli
```

If you're interested in contributing, you'll want to [install from master branch](https://github.com/mapbox/mapbox-cli-py/blob/master/CONTRIBUTING.md#contributing).

## Setup

Use of the `mapbox` command line interface requires an access token.
Your token is shown on the [API access tokens](https://www.mapbox.com/studio/account/tokens/) page when you are logged in.
The token can be provided on the command line

```
$ mapbox --access-token MY_TOKEN ...
```

or as an environment variable named `MAPBOX_ACCESS_TOKEN`

```
$ export MAPBOX_ACCESS_TOKEN=MY_TOKEN
$ mapbox ...
```

## Usage

* [directions](#directions)
* [geocoding](#geocoding)
* [mapmatching](#mapmatching)
* [staticmap](#staticmap)
* [upload](#upload)
* [datasets](#datasets)
* [analytics](#analytics)

For any command that takes waypoints or features as an input you can either specify:

* Coordinate pair(s) of the form `"[0, 0]"` or `"0,0"` or `"0 0"`
* Sequence of GeoJSON features on `stdin`
* GeoJSON FeatureCollection on `stdin`
* Paths to GeoJSON file(s) containing either a single Feature or FeatureCollection.

Note that functions that accept points only, any non-point feature is filtered out.

### directions
```
Usage: mapbox directions [OPTIONS] FEATURES...

  The Mapbox Directions API will show you how to get where you're going.

  mapbox directions "[0, 0]" "[1, 1]"

  An access token is required.  See "mapbox --help".

Options:
  --profile [mapbox/driving|mapbox/driving-traffic|mapbox/walking|mapbox/cycling]
                                  Routing profile
  --alternatives / --no-alternatives
                                  Whether to try to return alternative routes
  --geometries [geojson|polyline|polyline6]
                                  Format of returned geometry
  --overview [full|simplified|False]
                                  Type of returned overview geometry
  --steps / --no-steps            Whether to return steps and turn-by-turn
                                  instructions
  --continue-straight / --no-continue-straight
                                  Whether to see the allowed direction of
                                  travel when departing the original waypoint
  --waypoint-snapping TEXT        Controls waypoint snapping
  --annotations TEXT              Additional metadata along the route
  --language TEXT                 Language of returned turn-by-turn
                                  instructions
  -o, --output TEXT               Save output to a file
  --help                          Show this message and exit.
```


### geocoding
```
Usage: mapbox geocoding [OPTIONS] [QUERY]

  This command returns places matching an address (forward mode) or places
  matching coordinates (reverse mode).

  In forward (the default) mode the query argument shall be an address such
  as '1600 pennsylvania ave nw'.

    $ mapbox geocoding '1600 pennsylvania ave nw'

  In reverse mode the query argument shall be a JSON encoded array of
  longitude and latitude (in that order) in decimal degrees.

    $ mapbox geocoding --reverse '[-77.4371, 37.5227]'

  An access token is required, see `mapbox --help`.

Options:
  --forward / --reverse           Perform a forward or reverse geocode.
                                  [default: forward]
  -i, --include                   Include HTTP headers in the output.
  --lat FLOAT                     Bias results toward this latitude (decimal
                                  degrees). --lon is also required.
  --lon FLOAT                     Bias results toward this longitude (decimal
                                  degrees). --lat is also required.
  -t, --place-type NAME           Restrict results to one or more place types.
  -o, --output TEXT               Save output to a file.
  -d, --dataset [mapbox.places|mapbox.places-permanent]
                                  Source dataset for geocoding, [default:
                                  mapbox.places]
  --country TEXT                  Restrict forward geocoding to specified
                                  country codes,comma-separated
  --bbox TEXT                     Restrict forward geocoding to specified
                                  bounding box,given in minX,minY,maxX,maxY
                                  coordinates.
  --features                      Return results as line-delimited GeoJSON
                                  Feature sequence, not a FeatureCollection
  --limit INTEGER                 Limit the number of returned features
  --help                          Show this message and exit.
```

### mapmatching
```
Usage: mapbox mapmatching [OPTIONS] FEATURES...

  Mapbox Map Matching API lets you use snap your GPS traces to the
  OpenStreetMap road and path network.

        $ mapbox mapmatching trace.geojson

  An access token is required, see `mapbox --help`.

Options:
  --gps-precision INTEGER         Assumed precision of tracking device
                                  (default 4 meters)
  --profile [mapbox.driving|mapbox.cycling|mapbox.walking]
                                  Mapbox profile id
  --help                          Show this message and exit.
```

### staticmap
```
Usage: mapbox staticmap [OPTIONS] MAPID OUTPUT

  Generate static map images from existing Mapbox map ids. Optionally
  overlay with geojson features.

    $ mapbox staticmap --features features.geojson mapbox.satellite out.png
    $ mapbox staticmap --lon -61.7 --lat 12.1 --zoom 12 mapbox.satellite
    out2.png

  An access token is required, see `mapbox --help`.

Options:
  --features TEXT              GeoJSON Features to render as overlay
  --lat FLOAT                  Latitude
  --lon FLOAT                  Longitude
  --zoom INTEGER               Zoom
  --size <INTEGER INTEGER>...  Image width and height in pixels
  --help                       Show this message and exit.
```


### upload
```
Usage: mapbox upload [OPTIONS] TILESET [INFILE]

  Upload data to Mapbox accounts. All endpoints require authentication.
  Uploaded data lands at https://www.mapbox.com/studio/tilesets/ and can be used in new
  or existing projects.

  You can specify the tileset id and input file

    $ mapbox upload username.data mydata.geojson

  Or specify just the tileset id and take an input file on stdin

    $ cat mydata.geojson | mapbox upload username.data

  The --name option defines the title as it appears in Studio and defaults
  to the last part of the tileset id, e.g. "data"

  Note that the tileset must start with your username. An access token with
  upload scope is required, see `mapbox --help`.

Options:
  --name TEXT  Name for the data upload
  --help       Show this message and exit.
```

### datasets
```
Usage: mapbox datasets [OPTIONS] COMMAND [ARGS]...

  Read and write GeoJSON from Mapbox-hosted datasets

  All endpoints require authentication. An access token with appropriate
  dataset scopes is required, see `mapbox --help`.

  Note that this API is currently a limited-access beta.

Options:
  --help  Show this message and exit.

Commands:
  create                 Create an empty dataset
  create-tileset         Generate a tileset from a dataset
  delete-dataset         Delete a dataset
  delete-feature         Delete a single feature from a dataset
  list                   List datasets
  list-features          List features in a dataset
  put-feature            Insert or update a single feature in a dataset
  read-dataset           Return information about a dataset
  read-feature           Read a single feature from a dataset
  update-dataset         Update information about a dataset
```

### datasets list
```
Usage: mapbox datasets list [OPTIONS]

  List datasets.

  Prints a list of objects describing datasets.

      $ mapbox datasets list

  All endpoints require authentication. An access token with `datasets:read`
  scope is required, see `mapbox --help`.

Options:
  -o, --output TEXT  Save output to a file
  --help             Show this message and exit.
```

### datasets create
```
Usage: mapbox datasets create [OPTIONS]

  Create a new dataset.

  Prints a JSON object containing the attributes of the new dataset.

      $ mapbox datasets create

  All endpoints require authentication. An access token with
  `datasets:write` scope is required, see `mapbox --help`.

Options:
  -n, --name TEXT         Name for the dataset
  -d, --description TEXT  Description for the dataset
  --help                  Show this message and exit.
```

### datasets read-dataset
```
Usage: mapbox datasets read-dataset [OPTIONS] DATASET

  Read the attributes of a dataset.

  Prints a JSON object containing the attributes of a dataset. The
  attributes: owner (a Mapbox account), id (dataset id), created (Unix
  timestamp), modified (timestamp), name (string), and description (string).

      $ mapbox datasets read-dataset dataset-id

  All endpoints require authentication. An access token with `datasets:read`
  scope is required, see `mapbox --help`.

Options:
  -o, --output TEXT  Save output to a file
  --help             Show this message and exit.
```

### datasets update-dataset
```
Usage: mapbox datasets update-dataset [OPTIONS] DATASET

  Update the name and description of a dataset.

  Prints a JSON object containing the updated dataset attributes.

      $ mapbox datasets update-dataset dataset-id

  All endpoints require authentication. An access token with
  `datasets:write` scope is required, see `mapbox --help`.

Options:
  -n, --name TEXT         Name for the dataset
  -d, --description TEXT  Description for the dataset
  --help                  Show this message and exit.
```

### datasets delete-dataset
```
Usage: mapbox datasets delete-dataset [OPTIONS] DATASET

  Delete a dataset.

      $ mapbox datasets delete-dataset dataset-id

  All endpoints require authentication. An access token with
  `datasets:write` scope is required, see `mapbox --help`.

Options:
  --help  Show this message and exit.
```

### datasets list-features
```
Usage: mapbox datasets list-features [OPTIONS] DATASET

  Get features of a dataset.

  Prints the features of the dataset as a GeoJSON feature collection.

      $ mapbox datasets list-features dataset-id

  All endpoints require authentication. An access token with `datasets:read`
  scope is required, see `mapbox --help`.

Options:
  -r, --reverse TEXT  Read features in reverse
  -s, --start TEXT    Feature id to begin reading from
  -l, --limit TEXT    Maximum number of features to return
  -o, --output TEXT   Save output to a file
  --help              Show this message and exit.
```

### datasets put-feature
```
Usage: mapbox datasets put-feature [OPTIONS] DATASET FID [FEATURE]

  Create or update a dataset feature.

  The semantics of HTTP PUT apply: if the dataset has no feature with the
  given `fid` a new feature will be created. Returns a GeoJSON
  representation of the new or updated feature.

      $ mapbox datasets put-feature dataset-id feature-id 'geojson-feature'

  All endpoints require authentication. An access token with
  `datasets:write` scope is required, see `mapbox --help`.

Options:
  -i, --input TEXT  File containing a feature to put
  --help            Show this message and exit.
```

### datasets read-feature
```
Usage: mapbox datasets read-feature [OPTIONS] DATASET FID

  Read a dataset feature.

  Prints a GeoJSON representation of the feature.

      $ mapbox datasets read-feature dataset-id feature-id

  All endpoints require authentication. An access token with `datasets:read`
  scope is required, see `mapbox --help`.

Options:
  -o, --output TEXT  Save output to a file
  --help             Show this message and exit.
```

### datasets delete-feature
```
Usage: mapbox datasets delete-feature [OPTIONS] DATASET FID

  Delete a feature.

      $ mapbox datasets delete-feature dataset-id feature-id

  All endpoints require authentication. An access token with
  `datasets:write` scope is required, see `mapbox --help`.

Options:
  --help  Show this message and exit.
```

### datasets create-tileset
```
Usage: mapbox datasets create-tileset [OPTIONS] DATASET TILESET

  Create a vector tileset from a dataset.

      $ mapbox datasets create-tileset dataset-id username.data

  Note that the tileset must start with your username and the dataset must
  be one that you own. To view processing status, visit
  https://www.mapbox.com/data/. You may not generate another tilesets from
  the same dataset until the first processing job has completed.

  All endpoints require authentication. An access token with `uploads:write`
  scope is required, see `mapbox --help`.

Options:
  -n, --name TEXT  Name for the tileset
  --help           Show this message and exit.
```

### analytics
```
Usage: mapbox analytics [OPTIONS]

  The Mapbox Analytics API returns the counts per day for a given resource
  and period.

  Long options:

  $ mapbox analytics --resource-type <resource type> --username <username>
  --id <resource id> --start <ISO-formatted start date> --end <ISO-formatted
  end date>

  Short options:

  $ mapbox analytics -t <resource type> -u <username> -i <resource id> -s
  <ISO-formatted start date> -e <ISO-formatted end date>

  An access token is required.  See "mapbox --help".

Options:
  -t, --resource-type [tokens|styles|accounts|tilesets]
                                  The resource being requested.  [required]
  -u, --username TEXT             The username for the account that owns the
                                  resource.  [required]
  -i, --id TEXT                   The id for the resource.
  -s, --start TEXT                The ISO-formatted start date.
  -e, --end TEXT                  The ISO-formatted end date.
  --help                          Show this message and exit.
```

## Alternative command syntax

When saving a fraction of a second matters you can call the mapboxcli module
directly instead of using the installed program.

```$ python -m mapboxcli --help```


CHANGES
0.8.0
-----
- Breaking change: remove the deprecated Distance API
- Breaking change: remove the deprecated Surface API
- Pin to mapbox-sdk-py version 0.16.1

0.7.1
-----
- Pin to mapbox-sdk-py version 0.14.0, the most recent compatible version

0.7.0
-----
- Breaking change: mapbox subcommands like 'upload' are no longer plugins. We
  are trading extensibility for speed (#95).
- New feature: when saving a fraction of a second matters, mapbox subcommands
  can now be executed more quickly via `python -m mapboxcli upload` (#95).
  The `mapbox` console script remains the best command for ordinary usage on
  all platforms.
- mapbox sdk 0.12.2

0.6.2 (2017-05-04)
------------------
- use mapbox sdk 0.12.1

0.6.1 (2017-04-10)
------------------
- Bugfix, ascii charaters in upload progressbar (see #98)

0.6.0 (2017-01-23)
------------------
- Breaking change: in mapbox-upload TILESET is now the first positional
  argument and source file the second, reversing the order set in 0.4.0 (#85).
- Add a uploads progress bar (#86).
- Skip staging for datasources in accessible S3 buckets (#49).
- Requirements: use mapbox SDK version >=0.11.

0.5.1 (2015-11-22)
------------------
- Use mapbox sdk 0.10.1

0.5.0 (2016-11-22)
------------------
- Output sequence of geojson --features from geocoding (#76)
- Ability to --limit number of geocoding results (#77)
- Removed the datasets batch-update-delete functionality (#79)
- Added patch mode functionality for uploads (#80)

0.4.0 (2016-06-08)
------------------
- Add support for geocoder bounding box filter (#73)
- mapbox upload arguments in INPUT OUTPUT order, takes input on stdin (#68)

0.3.1 (2016-02-23)
------------------
- Bug fix for staticmap feature input (#60)

0.3.0 (2016-02-22)
------------------
- Testing and Documentation improvements (#38, #47, #54)
- Geocoding options to specify country code and dataset (#55)

0.2.0 (2015-12-18)
------------------
- Addition of mapbox-mapmatching command (#17).
- Access tokens can be set in a mapbox.ini config file (#1). See
  `mapbox --help` for directions.
- Addition of mapbox-config command (#30).
- The mapbox-upload infile arg type has been changed to `click.File` (#29).
- This release requires a patched distribution of `boto3`. Either follow the
  two step installation process in the README or satisfy the requirement in
  a single command using pip's `--constraint` option and the mapbox-cli-py
  constraints.txt file (#22).

0.1.0 (2015-12-08)
------------------
- Initial release consists of a `mapbox` command and the following
  sub-commands: directions, distance, geocoding, staticmap, surface, upload.
  Each corresponds to one of the Mapbox APIs described at
  https://www.mapbox.com/developers/api/.


mbx/scripts/geocoding.py
import json
import logging
import re

import click
import mapbox

from mbx.compat import map


class MapboxCLIException(click.ClickException):
    pass


def iter_query(query):
    """Accept a filename, stream, or string.
    Returns an iterator over lines of the query."""
    try:
        itr = click.open_file(query).readlines()
    except IOError:
        itr = [query]
    return itr


def coords_from_query(query):
    """Transform a query line into a (lng, lat) pair of coordinates."""
    try:
        coords = json.loads(query)
    except ValueError:
        vals = re.split(r"\,*\s*", query.strip())
        coords = [float(v) for v in vals]
    return tuple(coords[:2])


def echo_headers(headers, file=None):
    """Echo headers, sorted."""
    for k, v in sorted(headers.items()):
        click.echo("{0}: {1}".format(k.title(), v), file=file)
    click.echo(file=file)


@click.command(short_help="Geocode an address or coordinates.")
@click.argument('query', default='-', required=False)
@click.option(
    '--forward/--reverse',
    default=True,
    help="Perform a forward or reverse geocode. [default: forward]")
@click.option('--include', '-i', 'include_headers',
              is_flag=True, default=False,
              help="Include HTTP headers in the output.")
@click.option(
    '--lat', type=float, default=None,
    help="Bias results toward this latitude (decimal degrees). --lon "
         "is also required.")
@click.option(
    '--lon', type=float, default=None,
    help="Bias results toward this longitude (decimal degrees). --lat "
         "is also required.")
@click.option(
    '--place-type', '-t', multiple=True, metavar='NAME', default=None,
    help="Restrict results to one or more of these place types: {0}.".format(
        sorted(mapbox.Geocoder().place_types.keys())))
@click.option('--output', '-o', default='-', help="Save output to a file.")
@click.pass_context
def geocoding(ctx, query, forward, include_headers, lat, lon, place_type, output):
    """This command returns places matching an address (forward mode) or
    places matching coordinates (reverse mode).

    In forward (the default) mode the query argument shall be an address
    such as '1600 pennsylvania ave nw'.

      $ mbx geocoding '1600 pennsylvania ave nw'

    In reverse mode the query argument shall be a JSON encoded array
    of longitude and latitude (in that order) in decimal degrees.

      $ mbx geocoding --reverse '[-77.4371, 37.5227]'

    An access token is required, see `mbx --help`.
    """
    verbosity = (ctx.obj and ctx.obj.get('verbosity')) or 2
    logger = logging.getLogger('mapbox')

    access_token = (ctx.obj and ctx.obj.get('access_token')) or None
    stdout = click.open_file(output, 'w')

    geocoder = mapbox.Geocoder(access_token=access_token)

    if forward:
        for q in iter_query(query):
            resp = geocoder.forward(
                q, types=place_type, lat=lat, lon=lon)
            if include_headers:
                echo_headers(resp.headers, file=stdout)
            if resp.status_code == 200:
                click.echo(resp.text, file=stdout)
            else:
                raise MapboxCLIException(resp.text.strip())
    else:
        for lon, lat in map(coords_from_query, iter_query(query)):
            resp = geocoder.reverse(lon=lon, lat=lat, types=place_type)
            if include_headers:
                echo_headers(resp.headers, file=stdout)
            if resp.status_code == 200:
                click.echo(resp.text, file=stdout)
            else:
                raise MapboxCLIException(resp.text.strip())
